File: ChelaVm/src/PEFormat.cpp
#include <cstring>
#include "ChelaVm/PEFormat.hpp"
#include "ChelaVm/Module.hpp"

namespace ChelaVm
{
    // COFF header.
    struct CoffHeader
    {
        uint16_t machineType;
        uint16_t numSections;
        uint32_t timeStamp;
        uint32_t symbolTablePointer;
        uint32_t numSymbols;
        uint16_t optionalHeaderSize;
        uint16_t characteristics;

        void Read(ModuleReader &reader)
        {
            reader >> machineType >> numSections >> timeStamp;
            reader >> symbolTablePointer >> numSymbols;
            reader >> optionalHeaderSize >> characteristics;
        }
    };

    // COFF section header record.
    struct CoffSectionHeader
    {
        uint8_t name[8];
        uint32_t virtualSize;
        uint32_t virtualAddress;
        uint32_t rawDataSize;
        uint32_t rawDataPointer;
        uint32_t relocationsPointer;
        uint32_t lineNumberPointer;
        uint16_t numRelocations;
        uint16_t numLineNumbers;
        uint32_t characteristics;

        void Read(ModuleReader &reader)
        {
            reader.Read(name, 8);
            reader >> virtualSize >> virtualAddress >> rawDataSize;
            reader >> rawDataPointer >> relocationsPointer;
            reader >> lineNumberPointer >> numRelocations;
            reader >> numLineNumbers >> characteristics;
        }
    };
        
    size_t PEFormat_GetEmbeddedModule(ModuleReader &reader)
    {
        // Read the MZ signature.
        uint8_t signature[4];
        reader.Read(signature, 2);
        if(signature[0] != 'M' || signature[1] != 'Z')
            throw ModuleException("Invalid PE signature.");
            
        // Read the PE signature offset.
        uint32_t peOffset;
        reader.Seek(0x3c, SEEK_SET);
        reader >> peOffset;
        reader.Seek(peOffset, SEEK_SET);
        
        // Read the PE\0\0 signature.
        reader.Read(signature, 4);
        if(signature[0] != 'P' || signature[1] != 'E' ||
           signature[2] != 0 || signature[3] != 0)
           throw ModuleException("Unsupported MS DOS programs.");
           
        // Read the COFF header.
        CoffHeader header;
        header.Read(reader);
        
        // Ignore the optional header.
        reader.Skip(header.optionalHeaderSize);
        
        // Read the sections until hit '.cbm'
        CoffSectionHeader sectionHeader;
        for(int i = 0; i < header.numSections; ++i)
        {
            // Read the section header.
            sectionHeader.Read(reader);
            
            // Check if this is the section.
            if(!strncmp((const char*)sectionHeader.name, ".cbm", 8))
            {
                // Found the section with the embedded module.
                reader.Seek(sectionHeader.rawDataPointer, SEEK_SET);
                return sectionHeader.rawDataSize;
            }
        }

        // Could't find an embedded module.
        throw ModuleException("Couldn't find embedded module.");        
    }
}

File: ChelaCompiler/Module/PEFormat.cs
using System.Text;

namespace Chela.Compiler.Module
{
    public static class PEFormat
    {
        // COFF header.
        public struct CoffHeader
        {
            public ushort machineType;
            public ushort numSections;
            public uint timeStamp;
            public uint symbolTablePointer;
            public uint numSymbols;
            public ushort optionalHeaderSize;
            public ushort characteristics;

            public void Read(ModuleReader reader)
            {
                reader.Read(out machineType);
                reader.Read(out numSections);
                reader.Read(out timeStamp);
                reader.Read(out symbolTablePointer);
                reader.Read(out numSymbols);
                reader.Read(out optionalHeaderSize);
                reader.Read(out characteristics);
            }
        };

        // COFF section header record.
        public struct CoffSectionHeader
        {
            public string name;
            public uint virtualSize;
            public uint virtualAddress;
            public uint rawDataSize;
            public uint rawDataPointer;
            public uint relocationsPointer;
            public uint lineNumberPointer;
            public ushort numRelocations;
            public ushort numLineNumbers;
            public uint characteristics;

            public void Read(ModuleReader reader)
            {
                name = reader.ReadString(8);
                reader.Read(out virtualSize);
                reader.Read(out virtualAddress);
                reader.Read(out rawDataSize);
                reader.Read(out rawDataPointer);
                reader.Read(out relocationsPointer);
                reader.Read(out lineNumberPointer);
                reader.Read(out numRelocations);
                reader.Read(out numLineNumbers);
                reader.Read(out characteristics);
            }
        }

        public static void GetEmbeddedModule(ModuleReader reader)
        {
            // Read the MZ signature.
            if(reader.ReadByte() != 'M' || reader.ReadByte() != 'Z')
                throw new ModuleException("Invalid PE signature.");

            // Read the PE offset.
            reader.SetPosition(0x3c);
            uint peOffset = reader.ReadUInt();

            // Read the PE\0\0 sinature
            reader.SetPosition(peOffset);
            if(reader.ReadByte() != 'P' || reader.ReadByte() != 'E' ||
                reader.ReadByte() != 0 || reader.ReadByte() != 0)
                throw new ModuleException("Unsupported MS DOS programs.");

            // Read the COFF header.
            CoffHeader header = new CoffHeader();
            header.Read(reader);

            // Ignore the optional header.
            reader.Skip(header.optionalHeaderSize);

            // Read the sections until finding the .cbm section.
            CoffSectionHeader sectionHeader = new CoffSectionHeader();
            for(int i = 0; i < header.numSections; ++i)
            {
                // Read the section header.
                sectionHeader.Read(reader);

                // If this is the .cbm section, done.
                if(sectionHeader.name == ".cbm")
                {
                    reader.SetPosition(sectionHeader.rawDataPointer);
                    return;
                }
            }

            // Couldn't find embedded module.
            throw new ModuleException("Couldn't find embedded Chela module.");
        }
    }
}

